File: AWS_GUI_Dev/BAWGUI/ViewModels/Folder.vb
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.IO

Public Class Folder
    Inherits ViewModelBase
    Public Sub New(filename, filetype, ByRef firstFile)
        _name = Path.GetFileName(filename)
        _fullName = filename
        'what If type Is nothing?
        _type = filetype
        'firstFile = Nothing
        _buildDirTree(filename, firstFile)
    End Sub
    Private _type As String
    Private _name As String
    Public Property Name As String
        Get
            Return _name
        End Get
        Set(ByVal value As String)
            _name = value
            OnPropertyChanged()
        End Set
    End Property
    Private _fullName As String
    Public Property FullName As String
        Get
            Return _fullName
        End Get
        Set(ByVal value As String)
            _fullName = value
            OnPropertyChanged()
        End Set
    End Property
    Private _subFolders As ObservableCollection(Of Folder)
    Public Property SubFolders As ObservableCollection(Of Folder)
        Get
            Return _subFolders
        End Get
        Set(ByVal value As ObservableCollection(Of Folder))
            _subFolders = value
            OnPropertyChanged()
        End Set
    End Property
    Private Sub _buildDirTree(filename, ByRef firstFile)
        If File.Exists(filename) Then
            If String.IsNullOrEmpty(firstFile) Then
                firstFile = filename
            End If
            'If firstFile IsNot Nothing Then
            '    firstFile = filename
            'End If
        ElseIf Directory.Exists(filename) Then
            _subFolders = New ObservableCollection(Of Folder)
            For Each path In Directory.GetDirectories(filename)
                _subFolders.Add(New Folder(path, _type, firstFile))
            Next
            For Each file In Directory.GetFiles(filename)
                'what If type Is nothing?
                If Path.GetExtension(file).Substring(1) = _type Then
                    _subFolders.Add(New Folder(file, _type, firstFile))
                End If
            Next
        Else
            Throw New Exception(vbCrLf + "Error: data file path """ + filename + """ does not exists!")
        End If
    End Sub
End Class

File: AWS_GUI_Dev/BAWGUI.Utilities/Folder.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;
using System.Collections.ObjectModel;
using System.ComponentModel;

namespace BAWGUI.Utilities
{
        public class Folder : ViewModelBase
        {
            public Folder(string filename, string filetype, ref string firstFile)
            {
                _name = Path.GetFileName(filename);
                _fullName = filename;
                // what If type Is nothing?
                _type = filetype;
                // firstFile = Nothing
                // _buildDirTree(filename, firstFile)
                //_findFirstFile(filename, filetype, ref firstFile);
            }
            private string _type;
            private string _name;
            public string Name
            {
                get
                {
                    return _name;
                }
                set
                {
                    _name = value;
                    OnPropertyChanged();
                }
            }
            private string _fullName;
            public string FullName
            {
                get
                {
                    return _fullName;
                }
                set
                {
                    _fullName = value;
                    OnPropertyChanged();
                }
            }
            private ObservableCollection<Folder> _subFolders;
            public ObservableCollection<Folder> SubFolders
            {
                get
                {
                    return _subFolders;
                }
                set
                {
                    _subFolders = value;
                    OnPropertyChanged();
                }
            }
            private void _buildDirTree(string filename, ref string firstFile)
            {
                if (File.Exists(filename))
                {
                    if (string.IsNullOrEmpty(firstFile))
                        firstFile = filename;
                }
                else if (Directory.Exists(filename))
                {
                    _subFolders = new ObservableCollection<Folder>();
                    foreach (var path in Directory.GetDirectories(filename))
                        _subFolders.Add(new Folder(path, _type, ref firstFile));
                    foreach (var file in Directory.GetFiles(filename))
                    {
                        // what If type Is nothing?
                        if (Path.GetExtension(file).Substring(1) == _type)
                        {
                            if (string.IsNullOrEmpty(firstFile))
                                _subFolders.Add(new Folder(file, _type, ref firstFile));
                        }
                    }
                }
                else
                    throw new Exception("\nError: data file path \"" + filename + "\" does not exists!");
            }
            //private void _findFirstFile(string filename, string filetype, ref string firstFile)
            //{
            //    if (File.Exists(filename))
            //    {
            //        if (Path.GetExtension(filename).Substring(1).ToLower() == filetype)
            //            firstFile = filename;
            //    }
            //    else if (Directory.Exists(filename))
            //    {
            //        foreach (var file in Directory.GetFiles(filename))
            //        {
            //            _findFirstFile(file, filetype, ref firstFile);
            //            if (!string.IsNullOrEmpty(firstFile))
            //                return;
            //        }
            //        foreach (var path in Directory.GetDirectories(filename))
            //        {
            //            _findFirstFile(path, filetype, ref firstFile);
            //            if (!string.IsNullOrEmpty(firstFile))
            //                return;
            //        }
            //    }
            //    else
            //        throw new Exception("\nError: data file path \"" + filename + "\" does not exists!");
            //}
        }
}
